pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -0,0 +1,289 @@
+	SecretKeyName      = "tls.key"
+)
+
+var (
+	log       = logger.Log
+	Converter *bindings.TypeConverter
+
+	isProtectedTrue = true
+	vpcRole         = "ccp_internal_operator"
+	readerPath      = "/"
+	readerRole      = "cluster_info_reader"
+
+	antreaClusterResourceType = "AntreaClusterControlPlane"
+	revision1                 = int64(1)
+)
+
+func init() {
+	Converter = bindings.NewTypeConverter()
+	Converter.SetMode(bindings.REST)
+}
+
+type NSXServiceAccountService struct {
+	common.Service
+	PrincipalIdentityStore   cache.Indexer
+	ClusterControlPlaneStore cache.Indexer
+}
+
+// InitializeNSXServiceAccount sync NSX resources
+func InitializeNSXServiceAccount(service common.Service) (*NSXServiceAccountService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(2)
+	nsxServiceAccountService := &NSXServiceAccountService{Service: service}
+	nsxServiceAccountService.PrincipalIdentityStore = cache.NewIndexer(KeyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: IndexFunc})
+	nsxServiceAccountService.ClusterControlPlaneStore = cache.NewIndexer(KeyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: IndexFunc})
+
+	go queryPrincipalIdentity(nsxServiceAccountService, &wg, fatalErrors)
+	go queryClusterControlPlane(nsxServiceAccountService, &wg, fatalErrors)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		return nsxServiceAccountService, err
+	}
+
+	return nsxServiceAccountService, nil
+}
+
+func (s *NSXServiceAccountService) CreateOrUpdateNSXServiceAccount(ctx context.Context, obj *v1alpha1.NSXServiceAccount) error {
+	clusterName := s.getClusterName(obj.Namespace, obj.Name)
+	normalizedClusterName := util.NormalizeId(clusterName)
+	// TODO: Support default project name in config as project
+	project := s.NSXConfig.CoeConfig.Cluster
+	vpcName := obj.Namespace
+	vpcPath := fmt.Sprintf("/orgs/default/projects/%s/vpcs/%s", util.NormalizeId(project), vpcName)
+
+	// generate certificate
+	subject := util.DefaultSubject
+	subject.CommonName = clusterName
+	cert, key, err := util.GenerateCertificate(&subject, util.DefaultValidDays)
+	if err != nil {
+		return err
+	}
+
+	// create PI
+	if _, exists, _ := s.PrincipalIdentityStore.GetByKey(normalizedClusterName); !exists {
+		pi, err := s.NSXClient.WithCertificateClient.Create(mpmodel.PrincipalIdentityWithCertificate{
+			IsProtected: &isProtectedTrue,
+			Name:        &normalizedClusterName,
+			NodeId:      &normalizedClusterName,
+			Role:        nil,
+			RolesForPaths: []mpmodel.RolesForPath{{
+				Path: &readerPath,
+				Roles: []mpmodel.Role{{
+					Role: &readerRole,
+				}},
+			}, {
+				Path: &vpcPath,
+				Roles: []mpmodel.Role{{
+					Role: &vpcRole,
+				}},
+			}},
+			CertificatePem: &cert,
+			Tags:           common.ConvertTagsToMPTags(s.buildBasicTags(obj)),
+		})
+		if err != nil {
+			return err
+		}
+		s.PrincipalIdentityStore.Add(pi)
```

Comment:
Unhandled error

Reply:
Do we need to handle error here? We believe that no error should occur for these calls.